packages/next/src/server/image-optimizer.ts

```diff
@@ -516,77 +524,91 @@ export async function optimizeImage({
+): Promise<ImageUpstream> {
+  try {
+    const mocked = createRequestResponseMocks({
+      url: href,
+      method: _req.method || 'GET',
+      headers: _req.headers,
+      socket: _req.socket,
+    })
 
-  if (isAbsolute) {
-    const upstreamRes = await fetch(href)
+    await handleRequest(mocked.req, mocked.res, nodeUrl.parse(href, true))
+    await mocked.res.hasStreamed
 
-    if (!upstreamRes.ok) {
-      Log.error('upstream image response failed for', href, upstreamRes.status)
+    if (!mocked.res.statusCode) {
+      Log.error('image response failed for', href, mocked.res.statusCode)
       throw new ImageError(
-        upstreamRes.status,
-        '"url" parameter is valid but upstream response is invalid'
+        mocked.res.statusCode,
+        '"url" parameter is valid but internal response is invalid'
       )
     }
 
-    upstreamBuffer = Buffer.from(await upstreamRes.arrayBuffer())
-    upstreamType =
-      detectContentType(upstreamBuffer) ||
-      upstreamRes.headers.get('Content-Type')
-    maxAge = getMaxAge(upstreamRes.headers.get('Cache-Control'))
-  } else {
-    try {
-      const mocked = createRequestResponseMocks({
-        url: href,
-        method: _req.method || 'GET',
-        headers: _req.headers,
-        socket: _req.socket,
-      })
-
-      await handleRequest(mocked.req, mocked.res, nodeUrl.parse(href, true))
-      await mocked.res.hasStreamed
-
-      if (!mocked.res.statusCode) {
-        Log.error('image response failed for', href, mocked.res.statusCode)
-        throw new ImageError(
-          mocked.res.statusCode,
-          '"url" parameter is valid but internal response is invalid'
-        )
-      }
-
-      upstreamBuffer = Buffer.concat(mocked.res.buffers)
-      upstreamType =
-        detectContentType(upstreamBuffer) ||
-        mocked.res.getHeader('Content-Type')
-      const cacheControl = mocked.res.getHeader('Cache-Control')
-      maxAge = cacheControl ? getMaxAge(cacheControl) : 0
-    } catch (err) {
-      Log.error('upstream image response failed for', href, err)
-      throw new ImageError(
-        500,
-        '"url" parameter is valid but upstream response is invalid'
-      )
-    }
+    const buffer = Buffer.concat(mocked.res.buffers)
+    const contentType = mocked.res.getHeader('Content-Type')
+    const cacheControl = mocked.res.getHeader('Cache-Control')
+    return { buffer, contentType, cacheControl }
+  } catch (err) {
+    Log.error('upstream image response failed for', href, err)
+    throw new ImageError(
+      500,
+      '"url" parameter is valid but upstream response is invalid'
+    )
   }
+}
 
-  if (upstreamType) {
-    upstreamType = upstreamType.toLowerCase().trim()
+export async function imageOptimizer(
+  imageUpstream: ImageUpstream,
+  paramsResult: {
+    href: ImageParamsResult['href']
+    width: ImageParamsResult['width']
+    quality: ImageParamsResult['quality']
+    mimeType: ImageParamsResult['mimeType']
+  },
```

Comment:
non-blocking suggestion

```suggestion
  paramsResult: Pick<
    ImageParamsResult,
    'href' | 'width' | 'quality' | 'mimeType'
  >
```